### PR TITLE
Model 1: LSTM

### PR DESCRIPTION
**Pull request by KvEijden**
Copy of imdb example with our abstract data

### CODE REVIEW

**Comment by daob** [comment]
Here how you can define custom metrics:
https://keras.rstudio.com/reference/metric_binary_accuracy.html

**Comment by vankesteren** [comment]
To try out the model, check out [my branch](https://github.com/vankesteren/systematic-reviews/tree/kees-model) which includes the data.

File: Model1.R [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+library(caret)
+
+
+#reading traindata schoot-lgmm-ptsd
+cat('Loading data..schoot-lgmm-ptsd.\n')
+train_data      <- read.csv(file = "./data/example_dataset_1/csv/schoot-lgmm-ptsd-traindata.csv",
+                            header= TRUE, sep= ",", stringsAsFactors = FALSE)
+train_data      <- train_data %>% select(title, abstract, included_ats, included_final)
+
+train_data$text <- paste0(train_data$abstract, train_data$title)
```
Comment (J535D165) [inline Model1.R]: ### Remark 1

Abstract before title?

### Remark 2

The behaviour of paste is a bit tricky with missing values:
``` r
> df <- data.frame(x=c("str",NA), y=c(NA, "str"))
> paste0(df$x, df$y)
[1] "strNA" "NAstr"
```

### Remark 3

Prevent mixture of tidyverse and base R. 

### Solution

``` r
library(tidyverse)
train_data <- read_csv("./data/example_dataset_1/csv/schoot-lgmm-ptsd-traindata.csv") %>%
  mutate(
    text = str_c(
      str_replace_na(title, ""),
      str_replace_na(abstract, ""),
      sep=" "
    )
  ) %>%
  select(text, included_ats, included_final)
```

File: Model1.R [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+
+
+##contrary to the idm example, we don't have a vocabulary yet
+##we use the keras tokenizer to create a vocabulary from the words in the abstracts and the titles
+##every word is assigned a integer token
+##the abstracts and titles are are transformed into sequences (vectors of tokens)
+library(keras)
+max_features <- 10000 
+tokenizer    <- text_tokenizer(num_words = max_features)
```
Comment (J535D165) [inline Model1.R]: `keras::text_tokenizer(num_words = max_features)`

File: Model1.R [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+#make ratio of included/not included more equal
+
+
+sequences<- texts_to_sequences(tokenizer, train_data$text)
+
+#we need sequences with fixed length. 
+#the 75% quantile of lengths of all tokenized text is used. Shorter sequences are padded with 0
+#this will favors titles because they are put in front of abstracts in the text strings
+cat('Pad sequences................\n')
+q75_text_len  <- quantile(sapply(sequences, length))[4]
```
Comment (J535D165) [inline Model1.R]: I prefer this:
``` r
q75_text_len  <- quantile(sapply(sequences, length), probs=0.75)
```

File: Model1.R [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+train_data$text <- paste0(train_data$abstract, train_data$title)
+train_data      <- train_data %>% select(text, included_ats, included_final)
+
+
+
+##contrary to the idm example, we don't have a vocabulary yet
+##we use the keras tokenizer to create a vocabulary from the words in the abstracts and the titles
+##every word is assigned a integer token
+##the abstracts and titles are are transformed into sequences (vectors of tokens)
+library(keras)
```
Comment (J535D165) [inline Model1.R]: Duplicate import

File: Model1.R [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+max_features <- 10000 
+tokenizer    <- text_tokenizer(num_words = max_features)
+tokenizer    %>% fit_text_tokenizer(train_data$text)
+
+##tokenizer$word_index gives mapping of words to their tokens (=rank)
+##there are many stop words 
+#make ratio of included/not included more equal
```
[the remark this answers is not in the document]
Comment (KvEijden) [reply Model1.R]: removed

File: Model1.R [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+sequences<- texts_to_sequences(tokenizer, train_data$text)
+
+#we need sequences with fixed length. 
+#the 75% quantile of lengths of all tokenized text is used. Shorter sequences are padded with 0
+#this will favors titles because they are put in front of abstracts in the text strings
+cat('Pad sequences................\n')
+q75_text_len  <- quantile(sapply(sequences, length))[4]
+
+sequences     <- pad_sequences(sequences, maxlen= q75_text_len, padding = "post", truncating = "post")
```
[the remark this answers is not in the document]
Comment (KvEijden) [reply Model1.R]: done in 0.3

File: Model1.R [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+
+
+#make train en test sets
+cat("Make train en test set with ratio 9:1 .......\n")
+no_samples    <- dim(sequences)[1]
+test_size     <- as.integer(no_samples * 0.10)
```
[the remark this answers is not in the document]
Comment (KvEijden) [reply Model1.R]: done in version 0.3

File: Model1.R [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+TESTSET       <- sample(1:no_samples, test_size)
+
+x_train       <- sequences[-TESTSET, ]
+y_train_ats   <- as.numeric(train_data$included_ats[-TESTSET])
```
[the remark this answers is not in the document]
Comment (KvEijden) [reply Model1.R]: done in version 0.3

File: Model1.R [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+model %>%
+    layer_embedding(input_dim = max_features, output_dim = 64) %>% 
+    layer_lstm(units = 64, dropout = 0.2, recurrent_dropout = 0.2) %>% 
+    layer_dense(units = 1, activation = 'sigmoid')
+
+# Try using different optimizers and different optimizer configs
+model %>% compile(
+    loss      = 'binary_crossentropy',
+    optimizer = 'adam',
+    metrics   = c('accuracy')
```
[the remark this answers is not in the document]
Comment (KvEijden) [reply Model1.R]: done in 0.3

File: Model1.R [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+cat("Train the model on abstracts as features and included after title screening as labels .......\n")
+model %>% fit(
+    x_train, y_train_ats,
+    batch_size       = 128,
+    epochs           = 15,
+    validation_split = 0.2
```
[the remark this answers is not in the document]
Comment (KvEijden) [reply Model1.R]: done in 0.3

File: Model1.R [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+)
+
+cat('Test loss on test set:', scores[[1]])
+cat('Test accuracy on test set', scores[[2]])
+
+prediction <- predict_classes(model, x_test)
+
+p          <- factor(x=prediction, levels= c("1", "0"))
+r          <- factor(x=y_test_ats, levels= c("1", "0"))
+conf_mod   <- confusionMatrix(p, r, mode= "everything", dnn= c("Prediction", "ATS"))
```
[the remark this answers is not in the document]
Comment (KvEijden) [reply Model1.R]: done in 0.3

File: Model1.R (hunk 2 above)
Comment (KvEijden) [reply Model1.R]: done in 0.3

File: Model1.R (hunk 3 above)
Comment (KvEijden) [reply Model1.R]: done in 0.3
also made it an parameter quantile_probs and moved setting of value to the top

File: Model1.R (hunk 4 above)
Comment (KvEijden) [reply Model1.R]: removed in 0.3

File: Model1.R [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+cat('x_train tensor: ', length(dim(x_train)), ' shape:', dim(x_train), '\n')
+cat('x_test  tensor: ', length(dim(x_test)), ' shape:', dim(x_test), '\n')
+
+
+cat("Building model ........\n")
+model <- keras_model_sequential()
+model %>%
+    layer_embedding(input_dim = max_features, output_dim = 64) %>% 
+    layer_lstm(units = 64, dropout = 0.2, recurrent_dropout = 0.2) %>% 
+    layer_dense(units = 1, activation = 'sigmoid')
```
[the remark this answers is not in the document]
Comment (KvEijden) [reply Model1.R]: done in 0.3

**Comment by J535D165** [comment]
**squash and merge** this one

**Pull request merged**